File: app/src/main/java/com/limelight/binding/input/virtual_controller/VirtualControllerConfigurationLoader.java
/**
 * Created by Karim Mreisi.
 */

package com.limelight.binding.input.virtual_controller;

import android.app.Activity;
import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.Configuration;
import android.util.DisplayMetrics;

import com.limelight.nvstream.input.ControllerPacket;
import com.limelight.preferences.PreferenceConfiguration;

import org.json.JSONException;
import org.json.JSONObject;

public class VirtualControllerConfigurationLoader {
    public static final String OSC_PREFERENCE = "OSC";

    private static int getPercent(
            int percent,
            int total) {
        return (int) (((float) total / (float) 100) * (float) percent);
    }

    // The default controls are specified using a grid of 128*72 cells at 16:9
    private static int screenScale(int units, int height) {
        return Math.round(((float) height / (float) 72) * (float) units);
    }
    private static int screenScaleReverse(int pixel, int height) {
        return Math.round((float) pixel / ((float) height / (float) 72));
    }

    private static DigitalPad createDigitalPad(
            final VirtualController controller,
            final Context context) {

        DigitalPad digitalPad = new DigitalPad(controller, context);
        digitalPad.addDigitalPadListener(new DigitalPad.DigitalPadListener() {
            @Override
            public void onDirectionChange(int direction) {
                VirtualController.ControllerInputContext inputContext =
                        controller.getControllerInputContext();

                if ((direction & DigitalPad.DIGITAL_PAD_DIRECTION_LEFT) != 0) {
                    inputContext.inputMap |= ControllerPacket.LEFT_FLAG;
                }
                else {
                    inputContext.inputMap &= ~ControllerPacket.LEFT_FLAG;
                }
                if ((direction & DigitalPad.DIGITAL_PAD_DIRECTION_RIGHT) != 0) {
                    inputContext.inputMap |= ControllerPacket.RIGHT_FLAG;
                }
                else {
                    inputContext.inputMap &= ~ControllerPacket.RIGHT_FLAG;
                }
                if ((direction & DigitalPad.DIGITAL_PAD_DIRECTION_UP) != 0) {
                    inputContext.inputMap |= ControllerPacket.UP_FLAG;
                }
                else {
                    inputContext.inputMap &= ~ControllerPacket.UP_FLAG;
                }
                if ((direction & DigitalPad.DIGITAL_PAD_DIRECTION_DOWN) != 0) {
                    inputContext.inputMap |= ControllerPacket.DOWN_FLAG;
                }
                else {
                    inputContext.inputMap &= ~ControllerPacket.DOWN_FLAG;
                }

                controller.sendControllerInputContext();
            }
        });

        return digitalPad;
    }

    private static DigitalButton createDigitalButton(
            final int elementId,
            final int keyShort,
            final int keyLong,
            final int layer,
            final String text,
            final int icon,
            final VirtualController controller,
            final Context context) {
        DigitalButton button = new DigitalButton(controller, elementId, layer, context);
        button.setText(text);
        button.setIcon(icon);

        button.addDigitalButtonListener(new DigitalButton.DigitalButtonListener() {
            @Override
            public void onClick() {
                VirtualController.ControllerInputContext inputContext =
                        controller.getControllerInputContext();
                inputContext.inputMap |= keyShort;

                controller.sendControllerInputContext();
            }

            @Override
            public void onLongClick() {
                VirtualController.ControllerInputContext inputContext =
                        controller.getControllerInputContext();
                inputContext.inputMap |= keyLong;

                controller.sendControllerInputContext();
            }

            @Override
            public void onRelease() {
                VirtualController.ControllerInputContext inputContext =
                        controller.getControllerInputContext();
                inputContext.inputMap &= ~keyShort;
                inputContext.inputMap &= ~keyLong;

                controller.sendControllerInputContext();
            }
        });

        return button;
    }

    private static DigitalButton createLeftTrigger(
            final int layer,
            final String text,
            final int icon,
            final VirtualController controller,
            final Context context) {
        LeftTrigger button = new LeftTrigger(controller, layer, context);
        button.setText(text);
        button.setIcon(icon);
        return button;
    }

    private static DigitalButton createRightTrigger(
            final int layer,
            final String text,
            final int icon,
            final VirtualController controller,
            final Context context) {
        RightTrigger button = new RightTrigger(controller, layer, context);
        button.setText(text);
        button.setIcon(icon);
        return button;
    }

    private static AnalogStick createLeftStick(
            final VirtualController controller,
            final Context context) {
        return new LeftAnalogStick(controller, context);
    }

    private static AnalogStick createRightStick(
            final VirtualController controller,
            final Context context) {
        return new RightAnalogStick(controller, context);
    }


    private static final int TRIGGER_L_BASE_X = 1;
    private static final int TRIGGER_R_BASE_X = 92;
    private static final int TRIGGER_DISTANCE = 23;
    private static final int TRIGGER_BASE_Y = 31;
    private static final int TRIGGER_WIDTH = 12;
    private static final int TRIGGER_HEIGHT = 9;

    // Face buttons are defined based on the Y button (button number 9)
    private static final int BUTTON_BASE_X = 106;
    private static final int BUTTON_BASE_Y = 1;
    private static final int BUTTON_SIZE = 10;

    private static final int DPAD_BASE_X = 4;
    private static final int DPAD_BASE_Y = 41;
    private static final int DPAD_SIZE = 30;

    private static final int ANALOG_L_BASE_X = 6;
    private static final int ANALOG_L_BASE_Y = 4;
    private static final int ANALOG_R_BASE_X = 98;
    private static final int ANALOG_R_BASE_Y = 42;
    private static final int ANALOG_SIZE = 26;

    private static final int L3_R3_BASE_Y = 60;

    private static final int START_X = 83;
    private static final int BACK_X = 34;
    private static final int START_BACK_Y = 64;
    private static final int START_BACK_WIDTH = 12;
    private static final int START_BACK_HEIGHT = 7;

    // Make the Guide Menu be in the center of START and BACK menu
    private static final int GUIDE_X = START_X-BACK_X;
    private static final int GUIDE_Y = START_BACK_Y;

    private static int screenHeight = 1080;
    private static int screenWidth = 1920;
    private static int baseYUnit = 0;

    public static void createDefaultLayout(final VirtualController controller, final Context context) {

        DisplayMetrics screen = context.getResources().getDisplayMetrics();
        PreferenceConfiguration config = PreferenceConfiguration.readPreferences(context);

        screenHeight = screen.heightPixels;
        baseYUnit = 0;
        if (config.halfHeightOscPortrait && context.getResources().getConfiguration().orientation == Configuration.ORIENTATION_PORTRAIT) {
            screenHeight /= 2;
            baseYUnit = 72;
        }

        // Displace controls on the right by this amount of pixels to account for different aspect ratios
        screenWidth = screen.widthPixels;
        int rightDisplacement = screenWidth - screenHeight * 16 / 9;

        // NOTE: Some of these getPercent() expressions seem like they can be combined
        // into a single call. Due to floating point rounding, this isn't actually possible.

        if (!config.onlyL3R3)
        {
            controller.addElement(createDigitalPad(controller, context),
                    screenScale(DPAD_BASE_X, screenHeight),
                    screenScale(DPAD_BASE_Y + baseYUnit, screenHeight),
                    screenScale(DPAD_SIZE, screenHeight),
                    screenScale(DPAD_SIZE, screenHeight)
            );

            controller.addElement(createDigitalButton(
                    VirtualControllerElement.EID_A,
                    !config.flipFaceButtons ? ControllerPacket.A_FLAG : ControllerPacket.B_FLAG, 0, 1,
                    !config.flipFaceButtons ? "A" : "B", -1, controller, context),
                    screenScale(BUTTON_BASE_X, screenHeight) + rightDisplacement,
                    screenScale(BUTTON_BASE_Y + 2 * BUTTON_SIZE + baseYUnit, screenHeight),
                    screenScale(BUTTON_SIZE, screenHeight),
                    screenScale(BUTTON_SIZE, screenHeight)
            );

            controller.addElement(createDigitalButton(
                    VirtualControllerElement.EID_B,
                    config.flipFaceButtons ? ControllerPacket.A_FLAG : ControllerPacket.B_FLAG, 0, 1,
                    config.flipFaceButtons ? "A" : "B", -1, controller, context),
                    screenScale(BUTTON_BASE_X + BUTTON_SIZE, screenHeight) + rightDisplacement,
                    screenScale(BUTTON_BASE_Y + BUTTON_SIZE + baseYUnit, screenHeight),
                    screenScale(BUTTON_SIZE, screenHeight),
                    screenScale(BUTTON_SIZE, screenHeight)
            );

            controller.addElement(createDigitalButton(
                    VirtualControllerElement.EID_X,
                    !config.flipFaceButtons ? ControllerPacket.X_FLAG : ControllerPacket.Y_FLAG, 0, 1,
                    !config.flipFaceButtons ? "X" : "Y", -1, controller, context),
                    screenScale(BUTTON_BASE_X - BUTTON_SIZE, screenHeight) + rightDisplacement,
                    screenScale(BUTTON_BASE_Y + BUTTON_SIZE + baseYUnit, screenHeight),
                    screenScale(BUTTON_SIZE, screenHeight),
                    screenScale(BUTTON_SIZE, screenHeight)
            );

            controller.addElement(createDigitalButton(
                    VirtualControllerElement.EID_Y,
                    config.flipFaceButtons ? ControllerPacket.X_FLAG : ControllerPacket.Y_FLAG, 0, 1,
                    config.flipFaceButtons ? "X" : "Y", -1, controller, context),
                    screenScale(BUTTON_BASE_X, screenHeight) + rightDisplacement,
                    screenScale(BUTTON_BASE_Y + baseYUnit, screenHeight),
                    screenScale(BUTTON_SIZE, screenHeight),
                    screenScale(BUTTON_SIZE, screenHeight)
            );

            controller.addElement(createLeftTrigger(
                    1, "LT", -1, controller, context),
                    screenScale(TRIGGER_L_BASE_X, screenHeight),
                    screenScale(TRIGGER_BASE_Y + baseYUnit, screenHeight),
                    screenScale(TRIGGER_WIDTH, screenHeight),
                    screenScale(TRIGGER_HEIGHT, screenHeight)
            );

            controller.addElement(createRightTrigger(
                    1, "RT", -1, controller, context),
                    screenScale(TRIGGER_R_BASE_X + TRIGGER_DISTANCE, screenHeight) + rightDisplacement,
                    screenScale(TRIGGER_BASE_Y + baseYUnit, screenHeight),
                    screenScale(TRIGGER_WIDTH, screenHeight),
                    screenScale(TRIGGER_HEIGHT, screenHeight)
            );

            controller.addElement(createDigitalButton(
                    VirtualControllerElement.EID_LB,
                    ControllerPacket.LB_FLAG, 0, 1, "LB", -1, controller, context),
                    screenScale(TRIGGER_L_BASE_X + TRIGGER_DISTANCE, screenHeight),
                    screenScale(TRIGGER_BASE_Y + baseYUnit, screenHeight),
                    screenScale(TRIGGER_WIDTH, screenHeight),
                    screenScale(TRIGGER_HEIGHT, screenHeight)
            );

            controller.addElement(createDigitalButton(
                    VirtualControllerElement.EID_RB,
                    ControllerPacket.RB_FLAG, 0, 1, "RB", -1, controller, context),
                    screenScale(TRIGGER_R_BASE_X, screenHeight) + rightDisplacement,
                    screenScale(TRIGGER_BASE_Y + baseYUnit, screenHeight),
                    screenScale(TRIGGER_WIDTH, screenHeight),
                    screenScale(TRIGGER_HEIGHT, screenHeight)
            );

            controller.addElement(createLeftStick(controller, context),
                    screenScale(ANALOG_L_BASE_X, screenHeight),
                    screenScale(ANALOG_L_BASE_Y + baseYUnit, screenHeight),
                    screenScale(ANALOG_SIZE, screenHeight),
                    screenScale(ANALOG_SIZE, screenHeight)
            );

            controller.addElement(createRightStick(controller, context),
                    screenScale(ANALOG_R_BASE_X, screenHeight) + rightDisplacement,
                    screenScale(ANALOG_R_BASE_Y + baseYUnit, screenHeight),
                    screenScale(ANALOG_SIZE, screenHeight),
                    screenScale(ANALOG_SIZE, screenHeight)
            );

            controller.addElement(createDigitalButton(
                    VirtualControllerElement.EID_BACK,
                    ControllerPacket.BACK_FLAG, 0, 2, "BACK", -1, controller, context),
                    screenScale(BACK_X, screenHeight),
                    screenScale(START_BACK_Y + baseYUnit, screenHeight),
                    screenScale(START_BACK_WIDTH, screenHeight),
                    screenScale(START_BACK_HEIGHT, screenHeight)
            );

            controller.addElement(createDigitalButton(
                    VirtualControllerElement.EID_START,
                    ControllerPacket.PLAY_FLAG, 0, 3, "START", -1, controller, context),
                    screenScale(START_X, screenHeight) + rightDisplacement,
                    screenScale(START_BACK_Y + baseYUnit, screenHeight),
                    screenScale(START_BACK_WIDTH, screenHeight),
                    screenScale(START_BACK_HEIGHT, screenHeight)
            );
        }
        else {
            controller.addElement(createDigitalButton(
                    VirtualControllerElement.EID_LSB,
                    ControllerPacket.LS_CLK_FLAG, 0, 1, "L3", -1, controller, context),
                    screenScale(TRIGGER_L_BASE_X, screenHeight),
                    screenScale(L3_R3_BASE_Y + baseYUnit, screenHeight),
                    screenScale(TRIGGER_WIDTH, screenHeight),
                    screenScale(TRIGGER_HEIGHT, screenHeight)
            );

            controller.addElement(createDigitalButton(
                    VirtualControllerElement.EID_RSB,
                    ControllerPacket.RS_CLK_FLAG, 0, 1, "R3", -1, controller, context),
                    screenScale(TRIGGER_R_BASE_X + TRIGGER_DISTANCE, screenHeight) + rightDisplacement,
                    screenScale(L3_R3_BASE_Y + baseYUnit, screenHeight),
                    screenScale(TRIGGER_WIDTH, screenHeight),
                    screenScale(TRIGGER_HEIGHT, screenHeight)
            );
        }

        if(config.showGuideButton){
            controller.addElement(createDigitalButton(VirtualControllerElement.EID_GDB,
                            ControllerPacket.SPECIAL_BUTTON_FLAG, 0, 1, "GUIDE", -1, controller, context),
                    screenScale(GUIDE_X, screenHeight)+ rightDisplacement,
                    screenScale(GUIDE_Y + baseYUnit, screenHeight),
                    screenScale(START_BACK_WIDTH, screenHeight),
                    screenScale(START_BACK_HEIGHT, screenHeight)
            );
        }

        controller.setOpacity(config.oscOpacity);
    }

    public static JSONObject prepareSave(JSONObject config) throws JSONException {
        // 靠右侧的右对齐
        int dPosX = config.optInt("LEFT") * 2 + config.optInt("WIDTH");
        if (dPosX < screenWidth) {
            config.put("R_SIDE", false);
            config.put("LEFT", screenScaleReverse(config.optInt("LEFT"), screenHeight));
        } else {
            config.put("R_SIDE", true);
            config.put("LEFT", screenScaleReverse(screenWidth - config.optInt("LEFT"), screenHeight));
        }
        config.put("TOP", screenScaleReverse(config.optInt("TOP"), screenHeight) - baseYUnit);
        config.put("WIDTH", screenScaleReverse(config.optInt("WIDTH"), screenHeight));
        config.put("HEIGHT", screenScaleReverse(config.optInt("HEIGHT"), screenHeight));
        return config;
    }

    public static JSONObject prepareLoad(JSONObject config) throws JSONException {
        if (config.optBoolean("R_SIDE")) {
            config.put("LEFT", screenWidth - screenScale(config.optInt("LEFT"), screenHeight));
        } else {
            config.put("LEFT", screenScale(config.optInt("LEFT"), screenHeight));
        }
        config.put("TOP", screenScale(config.optInt("TOP") + baseYUnit, screenHeight));
        config.put("WIDTH", screenScale(config.optInt("WIDTH"), screenHeight));
        config.put("HEIGHT", screenScale(config.optInt("HEIGHT"), screenHeight));
        return config;
    }

    public static void saveProfile(final VirtualController controller,
                                   final Context context) {
        SharedPreferences.Editor prefEditor = context.getSharedPreferences(OSC_PREFERENCE, Activity.MODE_PRIVATE).edit();

        for (VirtualControllerElement element : controller.getElements()) {
            String prefKey = ""+element.elementId;
            try {
                prefEditor.putString(prefKey, prepareSave(element.getConfiguration()).toString());
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }

        prefEditor.apply();
    }

    public static void loadFromPreferences(final VirtualController controller, final Context context) {
        SharedPreferences pref = context.getSharedPreferences(OSC_PREFERENCE, Activity.MODE_PRIVATE);

        for (VirtualControllerElement element : controller.getElements()) {
            String prefKey = ""+element.elementId;

            String jsonConfig = pref.getString(prefKey, null);
            if (jsonConfig != null) {
                try {
                    element.loadConfiguration(prepareLoad(new JSONObject(jsonConfig)));
                } catch (JSONException e) {
                    e.printStackTrace();

                    // Remove the corrupt element from the preferences
                    pref.edit().remove(prefKey).apply();
                }
            }
        }
    }
}

File: app/src/main/java/com/limelight/binding/input/virtual_controller/VirtualControllerConfigurationLoader.kt
/** Originally created by Karim Mreisi. */
package com.limelight.binding.input.virtual_controller

import android.content.Context
import android.content.res.Configuration
import com.limelight.nvstream.input.ControllerPacket
import com.limelight.preferences.PreferenceConfiguration
import com.limelight.binding.input.virtual_controller.VirtualControllerElement.Companion.EID_A
import com.limelight.binding.input.virtual_controller.VirtualControllerElement.Companion.EID_B
import com.limelight.binding.input.virtual_controller.VirtualControllerElement.Companion.EID_X
import com.limelight.binding.input.virtual_controller.VirtualControllerElement.Companion.EID_Y
import com.limelight.binding.input.virtual_controller.VirtualControllerElement.Companion.EID_LB
import com.limelight.binding.input.virtual_controller.VirtualControllerElement.Companion.EID_RB
import com.limelight.binding.input.virtual_controller.VirtualControllerElement.Companion.EID_BACK
import com.limelight.binding.input.virtual_controller.VirtualControllerElement.Companion.EID_START
import com.limelight.binding.input.virtual_controller.VirtualControllerElement.Companion.EID_LSB
import com.limelight.binding.input.virtual_controller.VirtualControllerElement.Companion.EID_RSB
import com.limelight.binding.input.virtual_controller.VirtualControllerElement.Companion.EID_GDB
import org.json.JSONException
import org.json.JSONObject
import kotlin.math.roundToInt

object VirtualControllerConfigurationLoader {
    const val OSC_PREFERENCE = "OSC"

    @JvmStatic
    fun createDefaultLayout(controller: VirtualController, context: Context, layoutWidth: Int, layoutHeight: Int) {
        val config = PreferenceConfiguration.readPreferences(context)
        val style = VirtualControllerLayout.fromPreference(config.oscLayout)
        controller.layoutStyle = style
        controller.onlyL3R3 = config.onlyL3R3
        val classic = style == VirtualControllerLayout.CLASSIC
        val half = config.halfHeightOscPortrait && context.resources.configuration.orientation == Configuration.ORIENTATION_PORTRAIT
        val available = if (half) layoutHeight / 2 else layoutHeight
        val height = if (classic) available else minOf(available, layoutWidth * 72 / 88)
        val scale = height / 72f
        controller.profileScale = scale
        controller.profileOffsetY = if (classic) { if (half) height else 0 } else layoutHeight - height
        controller.profileWidth = layoutWidth
        controller.profileHeight = layoutHeight
        val accent = when (style) {
            VirtualControllerLayout.XBOX -> 0xFF8ED269.toInt()
            VirtualControllerLayout.DS -> 0xFF83B7FF.toInt()
            VirtualControllerLayout.NS -> 0xFF56CEE5.toInt()
            else -> 0xF00000FF.toInt()
        }
        fun add(element: VirtualControllerElement, x: Int, y: Int, w: Int, h: Int, right: Boolean = false) {
            if (!classic) {
                val color = if (style == VirtualControllerLayout.NS && right) 0xFFFF7C83.toInt() else accent
                element.setColors(if (style == VirtualControllerLayout.NS) color else 0xFFE1E6ED.toInt(), color)
            }
            controller.addElement(element,
                (x * scale).roundToInt() + if (right) layoutWidth - (128 * scale).roundToInt() else 0,
                (y * scale).roundToInt() + controller.profileOffsetY,
                // Round shared boundaries once so adjacent hit targets never overlap by a pixel.
                if (classic) (w * scale).roundToInt() else ((x + w) * scale).roundToInt() - (x * scale).roundToInt(),
                if (classic) (h * scale).roundToInt() else ((y + h) * scale).roundToInt() - (y * scale).roundToInt())
        }
        fun button(id: Int, flag: Int, label: String, layer: Int = 100 + id): DigitalButton =
            DigitalButton(controller, id, layer, context).apply {
                setText(label)
                addDigitalButtonListener(object : DigitalButton.DigitalButtonListener {
                    override fun onClick() = controller.setButtonState(this, flag, true)
                    override fun onLongClick() {}
                    override fun onRelease() = controller.setButtonState(this, flag, false)
                })
            }
        if (!config.onlyL3R3) {
            val symmetric = style == VirtualControllerLayout.DS
            val dpad = DigitalPad(controller, context).apply {
                contentDescription = context.getString(com.limelight.R.string.osc_dpad)
                addDigitalPadListener { direction ->
                    val flags = intArrayOf(ControllerPacket.LEFT_FLAG, ControllerPacket.UP_FLAG, ControllerPacket.RIGHT_FLAG, ControllerPacket.DOWN_FLAG)
                    var bits = 0
                    flags.forEachIndexed { i, flag -> bits = if (direction and (1 shl i) != 0) bits or flag else bits and flag.inv() }
                    controller.setButtonState(this, bits, bits != 0)
                }
            }
            add(dpad, if (classic) 4 else 5, if (classic) 41 else if (symmetric) 24 else 48,
                if (classic) 30 else 22, if (classic) 30 else 22)
            add(LeftAnalogStick(controller, context), if (classic) 6 else 4,
                if (classic) 4 else if (symmetric) 46 else 21, 26, 26)
            add(RightAnalogStick(controller, context), 98, if (classic) 42 else 46, 26, 26, true)
            // IDs are physical positions. NS and the face-flip preference each swap labels and wire flags.
            val swapped = (style == VirtualControllerLayout.NS) xor config.flipFaceButtons
            val flags = if (swapped) intArrayOf(ControllerPacket.B_FLAG, ControllerPacket.A_FLAG, ControllerPacket.Y_FLAG, ControllerPacket.X_FLAG)
                else intArrayOf(ControllerPacket.A_FLAG, ControllerPacket.B_FLAG, ControllerPacket.X_FLAG, ControllerPacket.Y_FLAG)
            val labels = if (symmetric) {
                if (swapped) arrayOf("○", "×", "△", "□") else arrayOf("×", "○", "□", "△")
            } else if (swapped) arrayOf("B", "A", "Y", "X") else arrayOf("A", "B", "X", "Y")
            val xs = if (classic) intArrayOf(106, 116, 96, 106) else intArrayOf(107, 117, 97, 107)
            val ys = if (classic) intArrayOf(21, 11, 11, 1) else intArrayOf(36, 26, 26, 16)
            val ids = intArrayOf(EID_A, EID_B, EID_X, EID_Y)
            flags.indices.forEach { i ->
                val key = button(ids[i], flags[i], labels[i], 1)
                if (style == VirtualControllerLayout.XBOX || symmetric) {
                    key.labelColor = when (flags[i]) {
                        ControllerPacket.A_FLAG -> if (symmetric) 0xFF83B7FF.toInt() else 0xFF8ED269.toInt()
                        ControllerPacket.B_FLAG -> 0xFFFF8792.toInt()
                        ControllerPacket.X_FLAG -> if (symmetric) 0xFFCCA2EF.toInt() else 0xFF83B7FF.toInt()
                        else -> if (symmetric) 0xFF80DABD.toInt() else 0xFFF1D475.toInt()
                    }
                }
                add(key, xs[i], ys[i], 10, 10, true)
            }
            val labelsShoulder = when (style) {
                VirtualControllerLayout.DS -> arrayOf("L2", "L1", "R1", "R2")
                VirtualControllerLayout.NS -> arrayOf("ZL", "L", "R", "ZR")
                else -> arrayOf("LT", "LB", "RB", "RT")
            }
            add(LeftTrigger(controller, if (classic) 1 else 2, context).apply { setText(labelsShoulder[0]) },
                if (classic) 1 else 3, if (classic) 31 else 6, 12, 9)
            add(button(EID_LB, ControllerPacket.LB_FLAG, labelsShoulder[1], if (classic) 1 else 3),
                if (classic) 24 else 17, if (classic) 31 else 6, 12, 9)
            add(button(EID_RB, ControllerPacket.RB_FLAG, labelsShoulder[2], if (classic) 1 else 4),
                if (classic) 92 else 99, if (classic) 31 else 6, 12, 9, true)
            add(RightTrigger(controller, if (classic) 1 else 5, context).apply { setText(labelsShoulder[3]) },
                if (classic) 115 else 113, if (classic) 31 else 6, 12, 9, true)
            val back = when (style) { VirtualControllerLayout.NS -> "−"; VirtualControllerLayout.DS -> "⋯"; else -> "BACK" }
            val start = when (style) { VirtualControllerLayout.NS -> "+"; VirtualControllerLayout.DS -> "≡"; else -> "START" }
            add(button(EID_BACK, ControllerPacket.BACK_FLAG, back, 6), if (classic) 34 else 30, if (classic) 64 else 60, 12, if (classic) 7 else 10)
            add(button(EID_START, ControllerPacket.PLAY_FLAG, start, 7), if (classic) 83 else 86, if (classic) 64 else 60, 12, if (classic) 7 else 10, true)
        }
        if (config.onlyL3R3 || !classic) {
            add(button(EID_LSB, ControllerPacket.LS_CLK_FLAG, "L3"), if (classic) 1 else 30, if (classic) 60 else 48, if (classic) 12 else 10, if (classic) 9 else 10)
            add(button(EID_RSB, ControllerPacket.RS_CLK_FLAG, "R3"), if (classic) 115 else 88, if (classic) 60 else 48, if (classic) 12 else 10, if (classic) 9 else 10, true)
        }
        if (config.showGuideButton) {
            val guide = button(EID_GDB, ControllerPacket.SPECIAL_BUTTON_FLAG, if (classic) "GUIDE" else "⌂")
            if (classic) add(guide, 49, 64, 12, 7, true)
            else {
                guide.setColors(0xFFE1E6ED.toInt(), accent)
                controller.addElement(guide, (layoutWidth - 10 * scale).roundToInt() / 2,
                    controller.profileOffsetY + ((if (layoutWidth / scale >= 100) 61 else 35) * scale).roundToInt(), (10 * scale).roundToInt(), (10 * scale).roundToInt())
            }
        }
        controller.setOpacity(config.oscOpacity)
    }

    internal fun profileKey(controller: VirtualController, id: Int): String =
        if (controller.layoutStyle == VirtualControllerLayout.CLASSIC) id.toString()
        else "${controller.layoutStyle.preferenceValue}.${if (controller.onlyL3R3) "minimal" else "full"}.$id"

    @JvmStatic
    fun saveProfile(controller: VirtualController, context: Context) {
        val editor = context.getSharedPreferences(OSC_PREFERENCE, Context.MODE_PRIVATE).edit()
        controller.elements.forEach { element ->
            val json = element.configuration
            val right = json.getInt("LEFT") * 2 + json.getInt("WIDTH") >= controller.profileWidth
            val x = json.getInt("LEFT")
            fun normalized(value: Int): Number =
                if (controller.layoutStyle == VirtualControllerLayout.CLASSIC) (value / controller.profileScale).roundToInt()
                else value.toDouble() / controller.profileScale
            json.put("R_SIDE", right)
            json.put("LEFT", normalized(if (right) controller.profileWidth - x else x))
            json.put("TOP", normalized(json.getInt("TOP") - controller.profileOffsetY))
            json.put("WIDTH", normalized(json.getInt("WIDTH")))
            json.put("HEIGHT", normalized(json.getInt("HEIGHT")))
            editor.putString(profileKey(controller, element.elementId), json.toString())
        }
        editor.apply()
    }

    @JvmStatic
    fun loadFromPreferences(controller: VirtualController, context: Context) {
        val prefs = context.getSharedPreferences(OSC_PREFERENCE, Context.MODE_PRIVATE)
        controller.elements.forEach { element ->
            val key = profileKey(controller, element.elementId)
            val saved = prefs.getString(key, null) ?: return@forEach
            try {
                val json = JSONObject(saved)
                fun scaled(name: String) = (json.getDouble(name) * controller.profileScale).roundToInt()
                val w = scaled("WIDTH").coerceIn(1, controller.profileWidth)
                val h = scaled("HEIGHT").coerceIn(1, controller.profileHeight)
                val x = if (json.optBoolean("R_SIDE")) controller.profileWidth - scaled("LEFT") else scaled("LEFT")
                val y = scaled("TOP") + controller.profileOffsetY
                json.put("LEFT", x.coerceIn(0, controller.profileWidth - w))
                json.put("TOP", y.coerceIn(0, controller.profileHeight - h))
                json.put("WIDTH", w).put("HEIGHT", h)
                element.loadConfiguration(json)
            } catch (_: JSONException) { prefs.edit().remove(key).apply() }
        }
    }
}
